server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -424,144 +410,212 @@ public ClusterState execute(ClusterState currentState) {
             }
             // Updating cluster state
             ClusterState.Builder builder = ClusterState.builder(currentState);
-            Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata());
+            Metadata currentMetadata = currentState.metadata();
+            Metadata.Builder mdBuilder = Metadata.builder(currentMetadata);
             ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
             RoutingTable.Builder rtBuilder = RoutingTable.builder(currentState.routingTable());
             ImmutableOpenMap<ShardId, RestoreInProgress.ShardRestoreStatus> shards;
             Set<String> aliases = new HashSet<>();
 
-            if (indices.isEmpty() == false) {
-                // We have some indices to restore
-                ImmutableOpenMap.Builder<ShardId, RestoreInProgress.ShardRestoreStatus> shardsBuilder =
-                    ImmutableOpenMap.builder();
-                final Version minIndexCompatibilityVersion = currentState.nodes().getMaxNodeVersion()
-                    .minimumIndexCompatibilityVersion();
-                for (Map.Entry<String, String> indexEntry : indices.entrySet()) {
-                    String index = indexEntry.getValue();
-                    ensureNotPartial(index);
-                    SnapshotRecoverySource recoverySource = new SnapshotRecoverySource(restoreUUID, snapshot,
-                        snapshotInfo.version(), repositoryData.resolveIndexId(index));
-                    IndexMetadata snapshotIndexMetadata = metadata.index(index);
-                    if (snapshotIndexMetadata == null) {
-                        throw new IndexNotFoundException(
-                            "Failed to find index '" + index + "' at the metadata of the current snapshot '" + snapshot + "'",
-                            index
-                        );
+            ImmutableOpenMap.Builder<ShardId, RestoreInProgress.ShardRestoreStatus> shardsBuilder =
+                ImmutableOpenMap.builder();
+            final Version minIndexCompatibilityVersion = currentState.nodes().getMaxNodeVersion()
+                .minimumIndexCompatibilityVersion();
+
+            HashMap<RelationName, List<String>> relationIndexUUIDs = new HashMap<>();
+            for (Map.Entry<String, String> indexEntry : indices.entrySet()) {
```

Comment:
Note the changes here are a bit easier to review if hiding whitespace diff - I removed the `indices.isEmpty()` which changed the indentation level.